File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
import string 

def solution(msg):
    answer = []
    
    # 초기 딕셔러니 생성
    dic = {}
    for num,st in enumerate(string.ascii_uppercase):
        dic[st] = num+1
        
    name = []
    cnt = 0
    # 문자열의 내용 반복문
    for n, s in enumerate(msg):
        if cnt :
            cnt -= 1
            continue
            
        else :
            # 첫번째 문자열에 대해서 사전에 존재하는지 최대 문자열 갯수 까지 늘려가면서 확인
            for i in range(1,len(msg)+1): # 1 부터 시작
                if msg[n:n+i] in dic:
                    cnt += 1
                    name.append(msg[n:n+i])
                    continue
                else :
                    dic[msg[n:n+i]] = len(dic)+1
                    answer.append(dic[msg[n:n+i-1]])
                    cnt -= 1
                    break

    answer.append(dic[name[-1]])

    return answer

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.java
import java.util.*;

class Solution {
    public int[] solution(String msg) {
        int[] answer = {};
        List<Integer> index = new ArrayList<>();
        Map<String, Integer> words = new HashMap<>();
        
        // A - Z 단어 사전 세팅
        for(int i = 0; i < 26; i++){
            words.put((char)('A' + i) + "", i+1);
        }
        
        StringBuilder sb = new StringBuilder();
        int len = msg.length();
        
        for(int i = 0; i < len; i++){
            String word = sb.toString() + msg.charAt(i);
            if (words.containsKey(word)){
                sb.append(msg.charAt(i));
                if ( i == len - 1 ){
                    index.add(words.get(sb.toString()));
                    break;
                }
                continue;
            }else{
                // 사전 추가
                words.put(word, words.size() + 1);
                index.add(words.get(sb.toString()));
                sb.delete(0, sb.length());
                i--;
            }
            
        }
        
        answer = index.stream().mapToInt(Integer::valueOf).toArray();
        return answer;
    }
}

